nixos.nix

```diff
@@ -436,26 +492,110 @@ in
+            # permissions than regular directories, whereas its parent
+            # should be owned by root and have regular permissions.
+            #
+            # This simply collects all the home directories and sets
+            # the appropriate permissions and ownership.
+            homeDirs =
+              foldl'
+                (state: dir:
+                  let
+                    defaultPerms = {
+                      mode = "0755";
+                      user = "root";
+                      group = "root";
+                    };
+                    homeDir = {
+                      directory = dir.home;
+                      dirPath = dir.home;
+                      home = null;
+                      mode = "0700";
+                      user = dir.user;
+                      group = users.${dir.user}.group;
+                      inherit defaultPerms;
+                      inherit (dir) persistentStoragePath;
+                    };
+                  in
+                  if dir.home != null then
+                    if !(elem homeDir state) then
+                      state ++ [ homeDir ]
+                    else
+                      state
+                  else
+                    state)
+                [ ]
+                explicitDirs;
+
+            # Generate entries for all parent directories of the
+            # argument directories, listed in the order they need to
+            # be created. The parent directories are assigned default
+            # permissions.
+            mkParentDirs = dirs:
+              let
+                # Create a new directory item from `dir`, the child
+                # directory item to inherit properties from and
+                # `path`, the parent directory path.
+                mkParent = dir: path: {
+                  directory = path;
+                  dirPath =
+                    if dir.home != null then
+                      concatPaths [ dir.home path ]
+                    else
+                      path;
+                  inherit (dir) persistentStoragePath home;
+                  inherit (dir.defaultPerms) user group mode;
+                };
+                # Create new directory items for all parent
+                # directories of a directory.
+                mkParents = dir:
+                  map (mkParent dir) (parentsOf dir.directory);
+              in
+              unique (flatten (map mkParents dirs));
+
+            # Parent directories of home folders. This is usually only
+            # /home, unless the user's home is in a non-standard
+            # location.
+            homeDirParents = mkParentDirs homeDirs;
+
+            # Parent directories of all explicitly listed directories.
+            parentDirs = mkParentDirs explicitDirs;
+
+            # All directories in the order they should be created.
+            allDirs = homeDirParents ++ homeDirs ++ parentDirs ++ explicitDirs;
```

Comment:
Maybe this list ought to be deduplicated by directory name?  Would potentially avoid setting/resetting directory ownership and permissions.

Reply:
Well, we're already running `unique` for all the different directory types individually and they shouldn't overlap. We're checking for duplicate items already. The issues I can see are: conflicting parent dir specifications for files and recursive dir or file specifications, but both should be handled with assertions, not deduplicated. Maybe I'm missing something; do you have an example?

Reply:
Hey; I know it's exceedingly, unbelievably, absurdly, laughably tardy, but:

> The issues I can see are: conflicting parent dir specifications for files and recursive dir or file specifications

Yes, this is precisely it.  Given a configuration like this:

```nix
environment.persistence = {
  "/abc" = {
    files = [
      { file = "/foo/bar"; parentDirectory.mode = "0755"; }
      { file = "/foo/lsgold"; parentDirectory.mode = "2700"; }
    ];
  };
};
```

We end up with a directory-creation script that looks like this:

```sh
#!/nix/store/5jw69mbaj5dg4l2bj58acg3gxywfszpj-bash-5.2p26/bin/bash
_status=0
trap "_status=1" ERR
/nix/store/ylachk42x2aa8x2hg4lx9gh70ks41igb-impermanence-create-directories '' '/abc' 'root' 'root' '0755' ''
/nix/store/ylachk42x2aa8x2hg4lx9gh70ks41igb-impermanence-create-directories '/abc' '/foo' 'root' 'root' '0755' ''
/nix/store/ylachk42x2aa8x2hg4lx9gh70ks41igb-impermanence-create-directories '/abc' '/foo' 'root' 'root' '2700' ''

exit $_status

```

This results in running `chown --reference=...`, `chmod --reference=...`, and various Bash file tests more than once.  Additionally, the mode `2700` is ignored, as `/abc/foo` already exists at the time of the third execution of the directory-creation script.

> [B]oth should be handled with assertions, not deduplicated.

Seems reasonable; I suggested deduplication as that would preserve the existing outcomes while cutting out unneeded operations.